### DISCUSSION TITLE
Can I get the largest substructure that contains a specific substructure?

### DISCUSSION THREAD

**Discussion by YoavShamir5** [rdkit/rdkit#7311] [Q&A] [answered]
Hello, I have a question regarding a unique use case. I have two molecules that share a substructure A. I would like to find their largest common substructure, lets call it substructure B, that contains substructure A within it. Is there a built-in way to do this in Rdkit? Thanks!

**Comment by vfscalfani** [comment]
Yes, I believe what you are referring to is the maximum common substructure. There is an example in the RDKit Getting Started Guide here:

https://www.rdkit.org/docs/GettingStartedInPython.html#maximum-common-substructure

Vin

**Reply by YoavShamir5** [reply]
Thanks for your reply! I am familiar with MCS, but am looking for something a bit more specific in my use case. Lets say that these are the two molecules:
![fig_a](https://github.com/rdkit/rdkit/assets/39048833/fcdbd681-29b8-4c39-bd1c-a3be27c70544)
MCS, when setting completeRingsOnly to True, will yield:
![fig_b](https://github.com/rdkit/rdkit/assets/39048833/cef8de7d-bdec-496a-b189-54dea4e4c541)
But I am looking for something else - I want to find the largest common substructure that contains (within the substructure) a nested substructure, in my case:
![acryl](https://github.com/rdkit/rdkit/assets/39048833/ec1bcca3-788d-4632-984d-742785409a3c)
So, in this case I want the output to be:
![res](https://github.com/rdkit/rdkit/assets/39048833/3cf499c5-6766-4e8e-83ac-b700f5847b57)

Here is the code for these compounds:

```
mol_a_smi = "O=C(C=C)C1CC1C2CC(C3CCC3C4C(C5CC5)CC4)CC2"
mol_a = Chem.MolFromSmiles(mol_a_smi)

mol_b_smi = "O=C(C=C)C1CC1CC2CC(C3CCC3C4C(C5CC5)CC4)CC2"
mol_b = Chem.MolFromSmiles(mol_b_smi)

mols = [mol_a,mol_b]

MolsToGridImage(mols, subImgSize=(300, 200), legends = ["A","B"])
```

**Reply by YoavShamir5** [reply]
I would still appreciate additional input on this query, thanks!

**Comment by DavidACosgrove** [comment, marked as answer]
I think the RASCAL MCES code can help you here.
```
from rdkit.Chem import rdRascalMCES
opts = rdRascalMCES.RascalOptions()
results = rdRascalMCES.FindMCES(mol1, mol2, opts)
for res in results:
    print(res.similarity, res.numFragments, res.smartsString)
```
which gives
```
0.9234615384615384 2 O=C(-C=C)-C1CC1.C1CC(-C2CCC2C2C(-C3CC3)-CC2)-CC
```
Mapping the SMARTS strings back onto the molecules gives
![image](https://github.com/rdkit/rdkit/assets/9198870/32e81e9d-fb94-4923-8699-b584feb8c5ee)
![image](https://github.com/rdkit/rdkit/assets/9198870/05d90579-b439-4b13-b3ad-55b9c27b6715)

where the blue bonds are the unmatched ones.  You would need to add some extra code to fragment the molecules based on bonds you don't want and then pick out the fragment that matches the substructure you care about.  I believe (though I can't prove it) that in a fragmented system like this, the largest common piece that has your substructure will always be the largest available such fragment.

See https://greglandrum.github.io/rdkit-blog/posts/2023-11-08-introducingrascalmces.html for more information about RASCAL.

**Reply by YoavShamir5** [reply]
I was not familiar with RascalMCES, looks super interesting and relevant, thanks a lot!
